common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/AbstractMicrosoftStsTokenResponseHandler.java

```diff
@@ -0,0 +1,175 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.providers.microsoft.microsoftsts;
+
+import static com.microsoft.identity.common.java.net.HttpConstants.HeaderField.XMS_CCS_REQUEST_ID;
+import static com.microsoft.identity.common.java.net.HttpConstants.HeaderField.XMS_CCS_REQUEST_SEQUENCE;
+import static com.microsoft.identity.common.java.net.HttpConstants.HeaderField.X_MS_CLITELEM;
+
+import com.google.gson.JsonParseException;
+import com.microsoft.identity.common.java.exception.ClientException;
+import com.microsoft.identity.common.java.flighting.CommonFlight;
+import com.microsoft.identity.common.java.flighting.CommonFlightsManager;
+import com.microsoft.identity.common.java.net.HttpResponse;
+import com.microsoft.identity.common.java.opentelemetry.AttributeName;
+import com.microsoft.identity.common.java.opentelemetry.SpanExtension;
+import com.microsoft.identity.common.java.providers.microsoft.MicrosoftTokenErrorResponse;
+import com.microsoft.identity.common.java.providers.oauth2.TokenErrorResponse;
+import com.microsoft.identity.common.java.providers.oauth2.TokenResult;
+import com.microsoft.identity.common.java.telemetry.CliTelemInfo;
+import com.microsoft.identity.common.java.util.HeaderSerializationUtil;
+import com.microsoft.identity.common.java.util.ObjectMapper;
+import com.microsoft.identity.common.java.util.ResultUtil;
+
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import lombok.NonNull;
+
+/**
+ * Abstract class for handling Microsoft STS token responses. The responses can from service can
+ * encrypted or decrypted or either can be expected depends on contract between client and service.
+ * Implementations of this class should handle the response accordingly. Standard response is
+ * implemented in {@link MicrosoftStsTokenResponseHandler}. For PRT protocol contracts, different
+ * implementation would be used that can handle encrypted and unencryted responses.
+ */
+public abstract class AbstractMicrosoftStsTokenResponseHandler {
+
+    private static final String TAG = AbstractMicrosoftStsTokenResponseHandler.class.getSimpleName();
+
+    /**
+     * Handle the token response from the service.
+     * @param response An {@link HttpResponse} from the service}
+     */
+    @NonNull
+    public TokenResult handleTokenResponse(@NonNull final HttpResponse response) throws ClientException {
+        final String methodTag = TAG + ":handleTokenResponse";
+
+        final MicrosoftStsTokenResponse tokenResponse;
+        final TokenErrorResponse tokenErrorResponse;
+
+        if (response.getStatusCode() >= HttpURLConnection.HTTP_BAD_REQUEST) {
```

Comment:
should we not be checking here if the status code is not 200 to determine success

Reply:
Depends on how we want to consider 3xx and other status codes like 0 or 1. Since this code is only moved around, am keeping changing success vs failure out of scope this PR.